07_errors/willshen8/syncStore.go

```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"strconv"
+	"sync"
+)
+
+type SyncStore struct {
+	syncStore sync.Map
+	nextID    uint32
+	mu        sync.Mutex
+}
+
+// NewMapStore initialise a new SyncStore
+func NewSyncStore() *SyncStore {
+	return &SyncStore{syncStore: sync.Map{}}
+}
+
+// CreatePuppy create a new puppy and store in mapStore.
+func (s *SyncStore) CreatePuppy(p *Puppy) (uint32, error) {
+	if i, err := strconv.Atoi(p.Value); err == nil {
+		if i < 0 {
+			return 0, &Error{
+				Message: "Puppy value can't be less than 0.",
+				Code:    1,
+			}
+		}
+	}
+	s.mu.Lock()
```

Comment:
The example provided by Dan for lab6 uses `s.mu.Lock()` followed directly by `defer s.mu.Unlock()` at the beginning of the each applicable function.  I believe using defer is good practice because then it doesn't matter where the function exits, the defer means it will always unlock before the function exits.

Reply:
Thanks, changed added.